{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "17bc83af-70f4-446e-b2c2-472db1dce577",
   "metadata": {},
   "source": [
    "Description: \n",
    "We integrated and analyzed the single-cell transcriptomes of 48 human pancreatic islets coming from 17 Healthy, 14 Pre-Diabetic and 17 Type 2 Diabetic cadaveric donors of matched sex, age, and ancestry. Our dataset consists of 245,878 high-quality cells, clustered into 14 cell types. We summarize the findings of the integrated cell type transcriptomes, as well as those of the re-integrated Alpha, Beta and Delta cell types.\n",
    "\n",
    "Reference:\n",
    "Single-cell decoding of human islet cell type-specific alterations in type 2 diabetes reveals converging genetic- and state-driven Î²-cell gene expression defects\n",
    "Khushdeep Bandesh, Efthymios Motakis, Siddhi Nargund, Romy Kursawe, Vijay Selvam, Redwan M Bhuiyan, Giray Naim Eryilmaz, Sai Nivedita Krishnan, Cassandra N. Spracklen, Duygu Ucar, Michael L. Stitzel\n",
    "bioRxiv 2025.01.17.633590; doi: https://doi.org/10.1101/2025.01.17.633590\n",
    "\n",
    "Data: https://datasets.cellxgene.cziscience.com/d45ff50f-90e1-4983-9388-c5b2ca1f2866.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cell-1",
   "metadata": {
    "attributes": {
     "classes": [
      "# Install libraries"
     ],
     "id": ""
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: scanpy in /opt/anaconda3/lib/python3.12/site-packages (1.10.1)\n",
      "Requirement already satisfied: anndata>=0.8 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.12.2)\n",
      "Requirement already satisfied: h5py>=3.1 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (3.11.0)\n",
      "Requirement already satisfied: joblib in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (1.4.2)\n",
      "Requirement already satisfied: legacy-api-wrap>=1.4 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (1.4.1)\n",
      "Requirement already satisfied: matplotlib>=3.6 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (3.7.5)\n",
      "Requirement already satisfied: natsort in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (8.4.0)\n",
      "Requirement already satisfied: networkx>=2.7 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (3.3)\n",
      "Requirement already satisfied: numba>=0.56 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.60.0)\n",
      "Requirement already satisfied: numpy>=1.23 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (1.26.4)\n",
      "Requirement already satisfied: packaging>=21.3 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (25.0)\n",
      "Requirement already satisfied: pandas>=1.5 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (2.2.2)\n",
      "Requirement already satisfied: patsy in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.5.6)\n",
      "Requirement already satisfied: pynndescent>=0.5 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.5.13)\n",
      "Requirement already satisfied: scikit-learn>=0.24 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (1.7.1)\n",
      "Requirement already satisfied: scipy>=1.8 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (1.13.1)\n",
      "Requirement already satisfied: seaborn>=0.13 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.13.2)\n",
      "Requirement already satisfied: session-info in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (1.0.1)\n",
      "Requirement already satisfied: statsmodels>=0.13 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.14.2)\n",
      "Requirement already satisfied: tqdm in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (4.66.5)\n",
      "Requirement already satisfied: umap-learn!=0.5.0,>=0.5 in /opt/anaconda3/lib/python3.12/site-packages (from scanpy) (0.5.9.post2)\n",
      "Requirement already satisfied: array-api-compat>=1.7.1 in /opt/anaconda3/lib/python3.12/site-packages (from anndata>=0.8->scanpy) (1.12.0)\n",
      "Requirement already satisfied: zarr!=3.0.*,>=2.18.7 in /opt/anaconda3/lib/python3.12/site-packages (from anndata>=0.8->scanpy) (3.1.2)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (1.4.4)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /opt/anaconda3/lib/python3.12/site-packages (from matplotlib>=3.6->scanpy) (2.9.0.post0)\n",
      "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /opt/anaconda3/lib/python3.12/site-packages (from numba>=0.56->scanpy) (0.43.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/lib/python3.12/site-packages (from pandas>=1.5->scanpy) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/lib/python3.12/site-packages (from pandas>=1.5->scanpy) (2023.3)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /opt/anaconda3/lib/python3.12/site-packages (from scikit-learn>=0.24->scanpy) (3.5.0)\n",
      "Requirement already satisfied: six in /opt/anaconda3/lib/python3.12/site-packages (from patsy->scanpy) (1.16.0)\n",
      "Requirement already satisfied: stdlib_list in /opt/anaconda3/lib/python3.12/site-packages (from session-info->scanpy) (0.11.1)\n",
      "Requirement already satisfied: donfig>=0.8 in /opt/anaconda3/lib/python3.12/site-packages (from zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy) (0.8.1.post1)\n",
      "Requirement already satisfied: numcodecs>=0.14 in /opt/anaconda3/lib/python3.12/site-packages (from numcodecs[crc32c]>=0.14->zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy) (0.16.2)\n",
      "Requirement already satisfied: typing-extensions>=4.9 in /opt/anaconda3/lib/python3.12/site-packages (from zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy) (4.11.0)\n",
      "Requirement already satisfied: pyyaml in /opt/anaconda3/lib/python3.12/site-packages (from donfig>=0.8->zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy) (6.0.1)\n",
      "Requirement already satisfied: crc32c>=2.7 in /opt/anaconda3/lib/python3.12/site-packages (from numcodecs[crc32c]>=0.14->zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy) (2.7.1)\n",
      "Requirement already satisfied: igraph in /opt/anaconda3/lib/python3.12/site-packages (0.11.9)\n",
      "Requirement already satisfied: texttable>=1.6.2 in /opt/anaconda3/lib/python3.12/site-packages (from igraph) (1.7.0)\n",
      "Requirement already satisfied: leidenalg in /opt/anaconda3/lib/python3.12/site-packages (0.10.2)\n",
      "Requirement already satisfied: igraph<0.12,>=0.10.0 in /opt/anaconda3/lib/python3.12/site-packages (from leidenalg) (0.11.9)\n",
      "Requirement already satisfied: texttable>=1.6.2 in /opt/anaconda3/lib/python3.12/site-packages (from igraph<0.12,>=0.10.0->leidenalg) (1.7.0)\n",
      "Requirement already satisfied: anndata in /opt/anaconda3/lib/python3.12/site-packages (0.12.2)\n",
      "Requirement already satisfied: array-api-compat>=1.7.1 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (1.12.0)\n",
      "Requirement already satisfied: h5py>=3.8 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (3.11.0)\n",
      "Requirement already satisfied: legacy-api-wrap in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (1.4.1)\n",
      "Requirement already satisfied: natsort in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (8.4.0)\n",
      "Requirement already satisfied: numpy>=1.26 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (1.26.4)\n",
      "Requirement already satisfied: packaging>=24.2 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (25.0)\n",
      "Requirement already satisfied: pandas!=2.1.2,>=2.1.0 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (2.2.2)\n",
      "Requirement already satisfied: scipy>=1.12 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (1.13.1)\n",
      "Requirement already satisfied: zarr!=3.0.*,>=2.18.7 in /opt/anaconda3/lib/python3.12/site-packages (from anndata) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/anaconda3/lib/python3.12/site-packages (from pandas!=2.1.2,>=2.1.0->anndata) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/lib/python3.12/site-packages (from pandas!=2.1.2,>=2.1.0->anndata) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/lib/python3.12/site-packages (from pandas!=2.1.2,>=2.1.0->anndata) (2023.3)\n",
      "Requirement already satisfied: donfig>=0.8 in /opt/anaconda3/lib/python3.12/site-packages (from zarr!=3.0.*,>=2.18.7->anndata) (0.8.1.post1)\n",
      "Requirement already satisfied: numcodecs>=0.14 in /opt/anaconda3/lib/python3.12/site-packages (from numcodecs[crc32c]>=0.14->zarr!=3.0.*,>=2.18.7->anndata) (0.16.2)\n",
      "Requirement already satisfied: typing-extensions>=4.9 in /opt/anaconda3/lib/python3.12/site-packages (from zarr!=3.0.*,>=2.18.7->anndata) (4.11.0)\n",
      "Requirement already satisfied: pyyaml in /opt/anaconda3/lib/python3.12/site-packages (from donfig>=0.8->zarr!=3.0.*,>=2.18.7->anndata) (6.0.1)\n",
      "Requirement already satisfied: crc32c>=2.7 in /opt/anaconda3/lib/python3.12/site-packages (from numcodecs[crc32c]>=0.14->zarr!=3.0.*,>=2.18.7->anndata) (2.7.1)\n",
      "Requirement already satisfied: six>=1.5 in /opt/anaconda3/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas!=2.1.2,>=2.1.0->anndata) (1.16.0)\n",
      "Requirement already satisfied: pandas in /opt/anaconda3/lib/python3.12/site-packages (2.2.2)\n",
      "Requirement already satisfied: numpy>=1.26.0 in /opt/anaconda3/lib/python3.12/site-packages (from pandas) (1.26.4)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/anaconda3/lib/python3.12/site-packages (from pandas) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/lib/python3.12/site-packages (from pandas) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/lib/python3.12/site-packages (from pandas) (2023.3)\n",
      "Requirement already satisfied: six>=1.5 in /opt/anaconda3/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install scanpy\n",
    "!pip install igraph\n",
    "!pip install leidenalg\n",
    "!pip install anndata\n",
    "!pip install pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0d4ab69c-5959-42ba-8d93-adc43e88cb63",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e52d75-8cfd-4f3a-8cbf-2baabcb7c0a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "url = 'https://datasets.cellxgene.cziscience.com/d45ff50f-90e1-4983-9388-c5b2ca1f2866.h5ad'\n",
    "response = requests.get(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "39f81685-2733-4203-9ea4-e000afb47ae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the file locally\n",
    "with open('d45ff50f-90e1-4983-9388-c5b2ca1f2866.h5ad', 'wb') as file:\n",
    "    file.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee37240e-fc10-4c7d-b215-d457d309a482",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_data = sc.read_h5ad(\"d45ff50f-90e1-4983-9388-c5b2ca1f2866.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45c33c6e-9ae0-4fb9-a3c3-82f56b10bdf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_h5ad = sc_data.copy()\n",
    "\n",
    "sc.pp.normalize_total(output_h5ad)\n",
    "\n",
    "sc.pp.log1p(output_h5ad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6382a136-0ff8-4d60-a094-1117cca6d0fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['vst.mean', 'vst.variance', 'vst.variance.expected',\n",
      "       'vst.variance.standardized', 'vst.variable', 'feature_is_filtered',\n",
      "       'feature_name', 'feature_reference', 'feature_biotype',\n",
      "       'feature_length', 'feature_type'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "#print(output_h5ad.var.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3ee035a9-e141-4c07-83a9-01ad6f1f0a84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000241860    ENSG00000241860\n",
      "ENSG00000286448    ENSG00000286448\n",
      "ENSG00000237491          LINC01409\n",
      "ENSG00000177757             FAM87B\n",
      "ENSG00000228794          LINC01128\n",
      "                        ...       \n",
      "ENSG00000261442    ENSG00000261442\n",
      "ENSG00000262815    ENSG00000262815\n",
      "ENSG00000267480          LINC02979\n",
      "ENSG00000236352    ENSG00000236352\n",
      "ENSG00000228459          LINC01546\n",
      "Name: feature_name, Length: 26936, dtype: category\n",
      "Categories (26896, object): ['A1BG', 'A1BG-AS1', 'A1CF', 'A2M', ..., 'ZYG11B', 'ZYX', 'ZZEF1', 'ZZZ3']\n"
     ]
    }
   ],
   "source": [
    "#print(output_h5ad.var['feature_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d46c3cad-6877-4c18-82f8-bdd32a8a0e68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KIF21A']\n"
     ]
    }
   ],
   "source": [
    "# Find genes containing 'KIF21A'\n",
    "#matches = [gene for gene in output_h5ad.var.feature_name if 'KIF21A' in gene]\n",
    "#print(matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07e6da7e-08ca-4ed8-985e-a9797f801c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "normal_h5ad = output_h5ad[output_h5ad.obs.disease == 'normal', :]\n",
    "T2D_h5ad = output_h5ad[output_h5ad.obs.disease == 'type 2 diabetes mellitus', :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4db99eea-3be9-4c47-8530-8c56709bb97e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = 'KIF21A'\n",
    "\n",
    "normal_expr = normal_h5ad[:, normal_h5ad.var.feature_name == gene_name].X.toarray().flatten()\n",
    "T2D_expr = T2D_h5ad[:, T2D_h5ad.var.feature_name == gene_name].X.toarray().flatten()\n",
    "\n",
    "normal_expr = list(filter(lambda x: x != 0, normal_expr))\n",
    "T2D_expr = list(filter(lambda x: x != 0, T2D_expr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca093035-fada-4a22-af56-fea20a2d6ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_length = max(len(normal_expr), len(T2D_expr))\n",
    "df = pd.DataFrame({\n",
    "    'normal': normal_expr + [None] * (max_length - len(normal_expr)),\n",
    "    'type 2 diabetes mellitus': T2D_expr + [None] * (max_length - len(T2D_expr)),\n",
    "})\n",
    "df.to_csv('sc_human_KIF21A_expr_disease.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f81e5ba-cc9c-4279-bab7-9558b07cc12c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({\n",
    "    'Kif21a': np.concatenate([normal_expr, T2D_expr]),\n",
    "    'Group': ['normal'] * len(normal_expr) + ['type 2 diabetes mellitus'] * len(T2D_expr)\n",
    "})\n",
    "\n",
    "# Create violin plot\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.violinplot(x='Group', y='Kif21a', data=df)\n",
    "plt.title(f'KIF21A Expression in human normal vs type 2 diabetes mellitus')\n",
    "plt.ylabel('KIF21A Gene Expression')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
